monkey/monkey_island/cc/repository/agent_machine_facade.py

```diff
@@ -0,0 +1,59 @@
+from functools import lru_cache
+
+from common.types import AgentID, MachineID
+from monkey_island.cc.models import Machine
+from monkey_island.cc.repository import IAgentRepository, IMachineRepository
+
+
+class AgentMachineFacade:
+    """
+    A simple facade for getting the Machine for an Agent
+    """
+
+    def __init__(self, agent_repository: IAgentRepository, machine_repository: IMachineRepository):
+        self._agent_repository = agent_repository
+        self._machine_repository = machine_repository
+
+    @lru_cache(maxsize=8192)
+    def get_machine_id_from_agent_id(self, agent_id: AgentID) -> MachineID:
+        """
+        Given an AgentID, get the MachineID of the machine the Agent ran on
+
+        :param agent_id: An AgentID
+        :return: The MachineID of the Machine that the Agent ran on
+        :raises UnknownRecordError: If no `Agent` with the specified `agent_id` could be found
+        :raises RetrievalError: If an error occurs while attempting to retrieve the `Agent`
+        """
+        return self._agent_repository.get_agent_by_id(agent_id).machine_id
+
+    def get_agent_machine(self, agent_id: AgentID) -> Machine:
+        """
+        Get the Machine for a given Agent, by AgentID
+
+        :param agent_id: AgentID of the Agent
+        :return: Machine that the Agent ran on
+        :raises UnknownRecordError: If no `Machine` for the specified `agent_id` could be found
+        :raises RetrievalError: If an error occurs while attempting to retrieve the `Machine`
+        """
+        machine_id = self.get_machine_id_from_agent_id(agent_id)
+        return self._machine_repository.get_machine_by_id(machine_id)
+
+    def update_agent_machine(self, agent_id: AgentID, machine: Machine):
```

Comment:
Why does this need the agent ID at all?

Reply:
It's being used to check that the machine is not being replaced for the agent (machine ID doesn't match). Do you think it's unnecessary?